p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,474 @@
+                    "All CREATE and UPDATE operations in document must follow the schema description".to_string(),
+                )),
+            }
+        })?;
+
+        Ok(())
+    }
+}
+
+/// A struct for building documents.
+#[derive(Debug)]
+pub struct DocumentBuilder {
+    /// An unsorted collection of operations which are associated with a particular document id.
+    operations: Vec<OperationWithMeta>,
+    /// Permissions for this document.
+    /// TODO: don't know what form this will take yet, this is a placeholder for now.
+    permissions: Option<Vec<Author>>,
+}
+
+impl DocumentBuilder {
+    /// Instantiate a new DocumentBuilder with a collection of operations.
+    pub fn new(mut operations: Vec<OperationWithMeta>) -> Self {
+        operations.sort_by(|a, b| a.operation_id().as_str().cmp(b.operation_id().as_str()));
+        Self {
+            operations,
+            permissions: None,
+        }
+    }
+
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Add permissions for this Document.
+    pub fn permissions(mut self, permissions: Vec<Author>) -> Self {
+        self.permissions = Some(permissions);
+        self
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
+        // find create message
+
+        let collect_create_operation: Vec<OperationWithMeta> =
+            self.operations_iter().filter(|op| op.is_create()).collect();
+
+        if collect_create_operation.len() > 1 {
+            return Err(DocumentBuilderError::MoreThanOneCreateOperation);
+        } else if collect_create_operation.is_empty() {
+            return Err(DocumentBuilderError::NoCreateOperation);
+        }
+
+        let create_operation = collect_create_operation.get(0).unwrap(); // unwrap as we know there is one item
+
+        // Get the author of this document from the create message
+        let author = create_operation.public_key();
+
+        // Get the document id from the create message
+        let document_id = create_operation.operation_id();
+
+        // Get the document id from the create message
+        let schema_hash = create_operation.schema();
+
+        // Normally we would get the schema string from the DB by it's hash
+        let schema = Schema::new(&schema_hash, DOCUMENT_SCHEMA)?;
+
+        // Instantiate graph and operations map
+        let mut graph = IncrementalTopo::new();
+        let mut operations = BTreeMap::new();
+
+        for op in self.operations() {
+            // Validate each operation against the document schema before continuing.
+            // NB. cddl crate not wasm supported yet.
```

Comment:
This is partly true, the https://crates.io/crates/cddl/0.9.0-beta.0 release has wasm support, but this can be part of another PR.

I would maybe use consistent `@TODO` commenting style, so its easier to find them when searching for todos.